extension/sigv4authextension/extension_test.go

```diff
@@ -0,0 +1,335 @@
+	tests := []struct {
+		name        	string
+		cfg    		*Config
+		shouldError 	bool
+	}{
+		{
+			name: "success_case",
+			cfg: &Config{
+				Region:     	"region",
+				Service: 	"service",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"regional",
+			},
+			shouldError: false,
+		},
+		{
+			name: "failure_case_before_RoundTripper()",
+			cfg: &Config{
+				Service: 	"aps",
+				RoleArn:     	"testrolearn",
+				STSEndpoint: 	"regional",
+			},
+			shouldError: true,
+		},
+	}
+
+	base := (http.RoundTripper)(http.DefaultTransport.(*http.Transport).Clone())
+	runtimeInfo := "runtimeInfo"
+
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			sa, err := newSigv4Extension(testcase.cfg, runtimeInfo, zap.NewNop())
+			if testcase.shouldError {
+				assert.Error(t, err)
+				assert.Nil(t, sa)
+				return
+			}
+
+			assert.NotNil(t, sa)
+			rt, err := RoundTripperWithCredentials(base, testcase.cfg, awsCreds, runtimeInfo)
+			assert.Nil(t, err)
+
+			si := rt.(*SigningRoundTripper)
+			assert.Equal(t, base, si.transport)
+			assert.Equal(t, testcase.cfg.Region, si.region)
+			assert.Equal(t, testcase.cfg.Service, si.service)
+			assert.Equal(t, runtimeInfo, si.runtimeInfo)
+		})
+	}
+}
+
+func TestPerRPCCredentials(t *testing.T) {
+	cfg := &Config{Region: "region", Service: "service", RoleArn: "rolearn", STSEndpoint: "regional"}
+	sa, err := newSigv4Extension(cfg, "", zap.NewNop())
+
+	rpc, err := sa.PerRPCCredentials()
+	assert.Nil(t, rpc)
+	assert.Error(t, err)
+}
+
+func TestGetAWSConfig(t *testing.T) {
+	tests := []struct {
+		name                	string
+		cfg          		*Config
+		STSRegionalEndpoint 	endpoints.STSRegionalEndpoint
+	}{
+		{
+			"success_case_without_sts_endpoint",
+			&Config{},
+			endpoints.UnsetSTSEndpoint,
+		},
+		{
+			"success_case_with_legacy_sts_endpoint",
+			&Config{STSEndpoint: "legacy"},
+			endpoints.LegacySTSEndpoint,
+		},
+		{
+			"success_case_with_regional_sts_endpoint",
+			&Config{STSEndpoint: "regional"},
+			endpoints.RegionalSTSEndpoint,
+		},
+	}
+	// run tests
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			config, err := getAWSConfig(testcase.cfg)
+			require.NoError(t, err, "Failed getAWSConfig")
+			require.NotNil(t, config)
+			require.Equal(t, config.STSRegionalEndpoint, testcase.STSRegionalEndpoint)
+		})
+	}
+}
+
+func TestGetCredsFromConfig(t *testing.T) {
+	tests := []struct {
+		name 	string
+		cfg 	*Config
+	}{
+		{
+			"success_case_without_role",
+			&Config{Region: "region", Service: "service"},
+		},
+		{
+			"success_case_with_role",
+			&Config{Region: "region", Service: "service", RoleArn: "rolearn"},
+		},
+		{
+			"success_case_with_role_and_regional_sts_endpoint",
+			&Config{Region: "region", Service: "service", RoleArn: "rolearn", STSEndpoint: "regional"},
+		},
+	}
+	// run tests
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			creds, err := getCredsFromConfig(testcase.cfg)
```

Comment:
Reminder to reason out test cases instead of only trying to hit code coverage - we expected this test to fail when no credentials on CI, see above. This does apply generally among all the tests, the objective is to test behavior, number of lines of code tested is a side effect.

Reply:
Going off of my comment above, should `getCredsFromConfig()` have different behavior than what it is right now? Currently, the only configuration that will cause it to return an error is a bad `STSEndpoint`. Again, please let me know if I am misunderstanding what you meant when you said "why this isn't failing due to missing creds?" or "we expected this test to fail when no credentials on CI"

Reply:
There are several ways of setting up creds like the access key / secret key env vars, a web identity token file, a request to the metadata server, but was hoping that if none of these is set up `getCredsFromConfig` would return an error. In these tests, since credentials aren't set up within CI, that's why we would expect them to be failing due to invalid credentials. I was hoping that at least `creds.Get()` would let us do this through this codepath

https://github.com/aws/aws-sdk-go/blob/93134df4fd5d1cdb72dd122a850b54d414792d71/aws/session/credentials.go#L140
https://github.com/aws/aws-sdk-go/blob/93134df4fd5d1cdb72dd122a850b54d414792d71/aws/session/credentials.go#L298

Reply:
Resolving this conversation after switching to aws sdk v2; testing retrieving credentials with a rolearn fails now, and we remove it due to being unable to test it (needs to fetch a token from AWS)